packages/lib/gpcircuits/circuits/proto-pod-gpc.circom

```diff
@@ -108,9 +109,61 @@ template ProtoPODGPC (
         entryValueHashes[i] <== entryProofSiblings[i][0];
     }
 
-    // Entry can be compared for equality (by hash) to another entry (by index).
-    // This can be disabled by comparing to self: entryEqualToOtherEntryIndex[i] = i
-    signal input entryEqualToOtherEntryByIndex[MAX_ENTRIES];
+    /*
+     * Virtual entries. These are entries derived from PODs'
+     * cryptographic data, e.g. the signers' public keys.
+     *
+     * This block pipes them in and forms signals to be fed into the
+     * entry constraint, tuple and list membership modules below, as well
+     * as those revealing their hashes where applicable.
+     */
+    
+    // Maximum number of virtual entries.
+    var MAX_VIRTUAL_ENTRIES = MAX_OBJECTS;
+
+    // Virtual entry value hashes deduced from cryptographic data. At
+    // the moment, this consists of hashed object signers' public
+    // keys.
+    signal virtualEntryValueHashes[MAX_VIRTUAL_ENTRIES];
+        for (var i = 0; i < MAX_OBJECTS; i++) {
+        virtualEntryValueHashes[i]
+            <== Poseidon(2)([objectSignerPubkeyAx[i], objectSignerPubkeyAy[i]]);
+    }
+    
+    // Boolean flags for virtual entry behaviour.
+    signal input virtualEntryIsValueHashRevealed /*MAX_VIRTUAL_ENTRIES packed bits*/;
```

Comment:
Give this a `Bits` suffix like the non-virtual one.

For the inputs which are bit-packed, there would be a benefit to combining them into a single input with normal entries.  I.e. virtualEntryIsValueHashRevealed could be in the higher bits of entryIsValueHashRevealedBits.  For non-packed inputs, there isn't an efficiency benefit, but it make sense if the inputs came in together, already using the combined array indexes which are used to refer to them below.  What do you think?

Reply:
> Give this a `Bits` suffix like the non-virtual one.
I thought the `Bits` suffix went on the bit array signal?

> For the inputs which are bit-packed, there would be a benefit to combining them into a single input with normal entries. I.e. virtualEntryIsValueHashRevealed could be in the higher bits of entryIsValueHashRevealedBits. For non-packed inputs, there isn't an efficiency benefit, but it make sense if the inputs came in together, already using the combined array indexes which are used to refer to them below. What do you think?
I'm not sure. I thought about doing this, but my reasoning for keeping it separate was to not have weird asymmetries in the input signal array sizes. Technically we already have this happening with `entryEqualToOtherEntryByIndex` though. Perhaps part of the reason for my hesitance is the somewhat ambiguous nomenclature (which you alluded to above): virtual entries vs... proper entries? Total entries for all entries?

Reply:
> I thought the Bits suffix went on the bit array signal?

It is.  `virtualEntryIsValueHashRevealed` is a bit array signal if I'm understanding it right (and if GitHub is attaching this comment to the right line).

> I'm not sure. I thought about doing this, but my reasoning for keeping it separate was to not have weird asymmetries in the input signal array sizes. Technically we already have this happening with entryEqualToOtherEntryByIndex though. Perhaps part of the reason for my hesitance is the somewhat ambiguous nomenclature (which you alluded to above): virtual entries vs... proper entries? Total entries for all entries?

The asymmetry is odd, though I think it makes sense so long as there's some understandable order to which parameters can take which types of entry-like inputs.  I'd want to avoid cases where we needed to either have gaps in indexes, or use different indexes for the same thing in different places.

Reply:
Oh, also I stand corrected on the Bits suffix.  You're right that the packed bits get the bare name, while Bits is used elsewhere for the unpacked array.